### PR TITLE
Restore Previous ExecutionContext Capture Behavior

### LINKED ISSUE #16759: Stream.ReadAsync continues on different thread
## Steps to Reproduce

1.  Use code: `await stream.ReadAsync (buf, 0, count);`

The above should continue on the same thread that called ReadAsync but seems to continue on a different thread.

awaiting a Task.Run works as expected.

Sample Xamarin.Mac project:

https://github.com/mrward/TestStreamReadAsync

Slightly more complicated console app with a custom synchronisation context.

https://github.com/mrward/TestStreamReadAsyncConsoleApp

Also affects VS Mac.

### Current Behavior

After the await the thread has changed.
awaiting a Task.Run works as expected.

### Expected Behavior

After the await the thread is the same as the one that called ReadAsync.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

Mono JIT compiler version 6.4.0.188 (2019-06/2c3aeaf3780 Tue Sep  3 15:10:54 EDT 2019)
Mono JIT compiler version 6.0.0.327 (2019-02/f8ea05bddcb Sat Aug 10 18:47:12 EDT 2019)

### PR DESCRIPTION
**Pull request by steveisok**
Restores previous ExecutionContext Capture behavior where we were copying a new SynchronizationContext as opposed to a reference to the existing one.

The old behavior caused SynchronizationContextAwaitTaskContinuations to not post to a SynchronizationContext.

Fixes #16759

Reverts a fix for #12421

### CODE REVIEW

File: mcs/class/corlib/Test/System.Threading/CancellationTokenSourceTest.cs [hunk 1]
```diff
@@ -496,33 +498,101 @@ public void DisposeRace ()
...
-						taskIsCancelled = true;
-					}
-				};
-			awaitAction ();
-			ct.Cancel (); // should not trigger SynchronizationContext.Post
-			Assert.IsTrue (taskIsCancelled);
-			SynchronizationContext.SetSynchronizationContext (mainContext);
+			AsyncPump.Run(async delegate {
+				var _tcs = new TaskCompletionSource<bool>();
+                var _cts = new CancellationTokenSource();
```
Comment (EgorBo) [inline mcs/class/corlib/Test/System.Threading/CancellationTokenSourceTest.cs]: formatting issues?

Comment (steveisok) [reply mcs/class/corlib/Test/System.Threading/CancellationTokenSourceTest.cs]: Seems so

**Comment by steveisok** [comment]
Here's an easy way to test this out... 

https://gist.github.com/steveisok/d1f0a72aa51643b205858ce688c8f22d
https://gist.github.com/steveisok/ece59b2704f649435ffffa43aeffe208

**Comment by steveisok** [comment]
@monojenkins build failed

**Comment by steveisok** [comment]
Unit test fails on wasm, needs skipped.

**Comment by steveisok** [comment]
@monojenkins build failed

**Comment by steveisok** [comment]
@monojenkins backport 2019-08

**Comment by steveisok** [comment]
@monojenkins backport 2019-06

**Pull request merged**